packages/service/ni_measurement_plugin_sdk_service/_internal/grpc_servicer.py

```diff
@@ -307,7 +320,8 @@ def GetMetadata(  # noqa: N802 - function name should be lowercase
         )
 
         measurement_signature = v2_measurement_service_pb2.MeasurementSignature(
-            configuration_parameters_message_type="ni.measurementlink.measurement.v2.MeasurementConfigurations",
+            configuration_parameters_message_type=self._service_info.service_class
+            + ".Configurations",
```

Comment:
This changes the message type advertised in GetMetadata. This message type is stored in .sfp files and generated LV/Python client code. Do previously saved files still work after this change?

Reply:
> I tried a previously saved .sfp that uses the SampleMeasurement, and it still worked fine. (I updated grpc_servicer.py in-place, and killed NI.Discovery.Service.)

https://github.com/ni/measurement-plugin-python/blob/main/examples/sample_measurement/SampleMeasurement.sfp contains a JSON-serialized copy of the configuration parameters:

`<Screen ClientId="{c67a0b30-2590-499f-9b92-dbd2ef02233c}" ConfigurationParameters="{&quot;@type&quot;:&quot;type.googleapis.com/ni.measurementlink.measurement.v1.MeasurementConfigurations&quot;,&quot;float in&quot;:0.06,&quot;double array in&quot;:[0.1,0.2,0.3],&quot;...`

If the `.sfp` still works, then I think this may be what's happening:
- InstrumentStudio ignores the type URL when deserializing the JSON configuration parameters.
- InstrumentStudio doesn't use this type URL when sending Measure requests; it calls GetMetadata and uses the message names returned in the measurement signature.

> But for the Python client code, yes, it would have to get regenerated. And presumably the generated LV client code, too.
>
> The reason I had to change it is because the new name is what's actually there in the descriptor_pool.
> 
> So what should I do?

First of all, #890 and #850 are separate issues. It sounds like you're trying to fix both, but you only referenced one in your PR description.

When I filed #850, I listed `Any.unpack` as a possible solution, but on further thought, it might not be one. If we use the configuration/output message names advertised in the measurement signature when we register messages with the Protobuf descriptor pool, then we are basically requiring all measurement plug-ins to advertise unique configuration/output message names. Besides #890, there may be other reasons for multiple services to advertise the same message names:
- Multiple service versions with the same service class
- Copy/pasted code (if the message name isn't derived from the service class)
- Deliberate reuse of shared message types

Also, it sounds like we should reconsider baking the configuration/output message names into the generated code. Perhaps the generated code should call `GetMetadata` and then use the message names returned in the measurement signature. However, we have to balance this against the desire for the generated code to be a snapshot of the measurement metadata & version at the time it was generated.

Reply:
> First of all, https://github.com/ni/measurement-plugin-python/issues/890 and https://github.com/ni/measurement-plugin-python/issues/850 are separate issues. It sounds like you're trying to fix both, but you only referenced one in your PR description.

I just discovered #890 today, and I've so far only fixed half of it here. If I fix the other half, too, I'll reference it.

> When I filed https://github.com/ni/measurement-plugin-python/issues/850, I listed Any.unpack as a possible solution, but on further thought, it might not be one.

Note that this is how the C# plugins behave. So is it okay there but not here?

> If we use the configuration/output message names advertised in the measurement signature when we register messages with the Protobuf descriptor pool, then we are basically requiring all measurement plug-ins to advertise unique configuration/output message names. ... Also, it sounds like we should reconsider baking the configuration/output message names into the generated code.

I think we _should_ use the advertised message names (and probably bake them into the generated code), but we need to fix it to advertise the _correct_ message names, if the user provided their own descriptor instead of using a generated one.

My concern for that, though, is if they change the contents of the message without versioning the message name. But that's more of a best practice that _they'd_ need to follow. (Though I'm now wondering if we should also auto-version the generated message names. Is it allowed to have multiple versions of the same Measurement running in parallel in the discovery service? If so then we'd need to version the message. And then we'd have to figure out how specific we need to be with the version. Just major version??)

Reply:
> > When I filed https://github.com/ni/measurement-plugin-python/issues/850, I listed Any.unpack as a possible solution, but on further thought, it might not be one.

>
> Note that this is how the C# plugins behave. So is it okay there but not here?

Writing C# plugins is not officially supported yet. Also, there is no C# client generator yet.

My concern about using `Any.Unpack` and the measurement signature message names is for the client generator, not the service. 

If the client generator registers the measurement signature message names in the descriptor pool, then it may not work properly when there is a name collision. It may be acceptable to allow collisions as long as the descriptors are identical, but a small change to the service's metadata will change the serialized descriptor, making it incompatible with clients for other services.

Most services only use their own messages, and services that use a generated client to communicate with other services should not conflict with that generated client (aside from the "all Python services use the same names" issue).

> I think we should use the advertised message names (and probably bake them into the generated code), but we need to fix it to advertise the correct message names, if the user provided their own descriptor instead of using a generated one.

If we do this I think we should include the service version in addition to the service class, or allow clients to include a version.

Reply:
So, it's sounding like "Python measurement services tolerate incorrect type_url" is actually a _feature_ rather than a bug. Which suggests we should just reject #850, and that #890 is little more than an internal naming inconsistency.

And perhaps we should even revert #889 that implemented #851, so there's no impact on the generated client code. (And instead we would need to make our C# services more tolerant of an incorrect/missing type_url.)

Reply:
That's one way to look at it, but I think it's a stretch to say that "multiple measurement clients share a single protobuf descriptor pool" implies "measurement clients should send incorrect type URLs."

If you follow the published APIs in ni-apis, you can write a measurement service without reusing any of NI's C#, Python, or LV code. I've done it before. :) If you do this, then it's natural to unpack the protobuf `Any` type in the normal way supported by your programming language's protobuf library.

So I think it is 100% reasonable for services to expect clients to send the type URL advertised in the service's measurement signature.

#850 goes beyond this and suggests that the Python service package should validate the type URL. I could be convinced that this lack of validation is indeed a feature. After all, Postel's Law: be conservative in what you send, be liberal in what you accept.

However, following Postel's Law also suggests that clients should set the type URL to the message type advertised in the measurement signature. I do not think intentionally sending the wrong type URL is a feature. Also, if clients don't use the message type in the measurement signature as the type URL, then what is it there for?

Also, all of this is completely orthogonal to whether clients use the message types in the measurement signature when they register messages in the protobuf descriptor pool. I think it is 100% possible for clients to follow Postel's Law but also do a custom mapping between what the service thinks the message type is and what protobuf thinks the message type is.

Reply:
Alright, I updated it to fix #890 instead, and modified the validation I added for #850 to just be a warning printed to STDERR.